edb/edgeql/compiler/astutils.py

```diff
@@ -121,16 +122,22 @@ def extend_path(expr: qlast.Expr, field: str) -> qlast.Path:
         return qlast.Path(steps=[expr, step])
 
 
+@dataclass
+class Params:
+    cast_params: List[
+        Tuple[qlast.TypeCast, Dict[Optional[str], str]]
+    ] = field(default_factory=list)
+    shaped_params: List[qlast.Parameter] = field(default_factory=list)
```

Comment:
Let's store the Shape here so we can report the source span over all of `$userId { id }`